File: Som/Topology/SimpleMatrixTopology.cs
using System;
using System.Collections.Generic;

namespace Som.Topology
{
    public class SimpleMatrixTopology : ITopology
    {
        public int RowCount { get; private set; }
        public int ColCount { get; private set; }
        public double WholeTopologyRadius { get; private set; }

        public SimpleMatrixTopology(int rowCount, int colCount)
        {
            ColCount = colCount;
            RowCount = rowCount;

            WholeTopologyRadius = Math.Max(ColCount, RowCount) / 2.0;
        }

        public int NeuronsCount { get { return RowCount * ColCount; } }

        public int GetNeuronNumber(Location location)
        {
            throw new NotImplementedException();
        }

        public Location GetNeuronLocation(int neuronNumber)
        {
            throw new NotImplementedException();
        }

        public IList<int> GetDirectlyConnectedNeurons(int neuronNumber)
        {
            var result = new List<int>();
            int upper = neuronNumber - ColCount;
            int down = neuronNumber + ColCount;
            int right = neuronNumber + 1;
            int left = neuronNumber - 1;

            if (upper >= 0) result.Add(upper);
            if (down < NeuronsCount) result.Add(down);
            if (right % ColCount != 0) result.Add(right);
            if ((left + 1) % ColCount != 0) result.Add(left);

            return result;
        }

        public Dictionary<int, double> GetNeuronsInRadius(int neuronNumber, double radius)
        {
            //return Slow_GetNeuronsInRadius(neuronNumber, radius);
            return Fast_SquarNeibo(neuronNumber, radius);

        }

        public bool Overlaps(int neuronA, int neuronB, double radius)
        {
            var nColPosA = neuronA % ColCount;
            var nRowPosA = neuronA / ColCount;
            var nColPosB = neuronB % ColCount;
            var nRowPosB = neuronB / ColCount;
            var colDiff = Math.Abs(nColPosA - nColPosB);
            var rowDiff = Math.Abs(nRowPosA - nRowPosB);
            return (colDiff*colDiff + rowDiff*rowDiff) < (radius*radius);
        }

        public Dictionary<int, double> Fast_SquarNeibo(int neuronNumber, double radius)
        {
            var result = new Dictionary<int, double>();

            var nColPos = neuronNumber % ColCount;
            var nRowPos = neuronNumber / ColCount;

            var r = (int)Math.Floor(radius + 1);
            
            //runs through rows
            var higherLevel = nColPos - r;
            if(higherLevel < 0)higherLevel = 0;
            var lowerLevel = higherLevel + 2*r;
            if (lowerLevel > RowCount) lowerLevel = RowCount;

            //runs through columns
            var left = nRowPos - r;
            if (left < 0) left = 0;
            var right = left + 2*r;
            if (right > ColCount) right = ColCount;


            for (int col = higherLevel; col < lowerLevel; col++)
            {
                for (int row = left; row < right; row++)
                {
                    var currDist = GetDistance(col, row, nColPos, nRowPos);
                    if (currDist < radius)
                    {
                        result.Add(row * ColCount + col, currDist);
                    }
                }
            }

            return result;
        }

        public Dictionary<int, double> Slow_GetNeuronsInRadius(int neuronNumber, double radius)
        {
            var neuronsCount = RowCount * ColCount;

            var neuronColPos = neuronNumber % ColCount;
            var neuronRowPos = neuronNumber / ColCount;

            var result = new Dictionary<int, double>();

            for (int col = 0; col < ColCount; col++)
            {
                for (int row = 0; row < RowCount; row++)
                {
                    var currDist = GetDistance(col, row, neuronColPos, neuronRowPos);
                    if (currDist < radius)
                    {
                        result.Add(row * ColCount + col, currDist);
                    }
                }
            }

            return result;
        }


        private double GetDistance(int col, int row, int neuronColPos, int neuronRowPos)
        {
            var xD = (neuronColPos - col);
            var yD = (neuronRowPos - row);
            return Math.Sqrt(xD * xD + yD * yD);
        }
    }
}

File: trunk/JavaSom/Topology/SimpleMatrixTopology.java
package Topology;

	public class SimpleMatrixTopology implements ITopology
	{
		private int privateRowCount;
		public final int getRowCount()
		{
			return privateRowCount;
		}
		private void setRowCount(int value)
		{
			privateRowCount = value;
		}
		private int privateColCount;
		public final int getColCount()
		{
			return privateColCount;
		}
		private void setColCount(int value)
		{
			privateColCount = value;
		}
		private double privateWholeTopologyRadius;
		public final double getWholeTopologyRadius()
		{
			return privateWholeTopologyRadius;
		}
		private void setWholeTopologyRadius(double value)
		{
			privateWholeTopologyRadius = value;
		}

		public SimpleMatrixTopology(int rowCount, int colCount)
		{
			setColCount(colCount);
			setRowCount(rowCount);

			setWholeTopologyRadius(Math.max(getColCount(), getRowCount()) / 2.0);
		}

		public final int getNeuronsCount()
		{
			return getRowCount() * getColCount();
		}

		public final java.util.List<Integer> GetDirectlyConnectedNeurons(int neuronNumber)
		{
			java.util.ArrayList<Integer> result = new java.util.ArrayList<Integer>();
			int upper = neuronNumber - getColCount();
			int down = neuronNumber + getColCount();
			int right = neuronNumber + 1;
			int left = neuronNumber - 1;

			if (upper >= 0)
			{
				result.add(upper);
			}
			if (down < getNeuronsCount())
			{
				result.add(down);
			}
			if (right % getColCount() != 0)
			{
				result.add(right);
			}
			if ((left + 1) % getColCount() != 0)
			{
				result.add(left);
			}

			return result;
		}

		public final java.util.HashMap<Integer, Double> GetNeuronsInRadius(int neuronNumber, double radius)
		{
			//return Slow_GetNeuronsInRadius(neuronNumber, radius);
			return Fast_SquarNeibo(neuronNumber, radius);

		}

		public final boolean Overlaps(int neuronA, int neuronB, double radius)
		{
			int nColPosA = neuronA % getColCount();
			int nRowPosA = neuronA / getColCount();
			int nColPosB = neuronB % getColCount();
			int nRowPosB = neuronB / getColCount();
			int colDiff = Math.abs(nColPosA - nColPosB);
			int rowDiff = Math.abs(nRowPosA - nRowPosB);
			return (colDiff*colDiff + rowDiff*rowDiff) < (radius*radius);
		}

		public final java.util.HashMap<Integer, Double> Fast_SquarNeibo(int neuronNumber, double radius)
		{
			java.util.HashMap<Integer, Double> result = new java.util.HashMap<Integer, Double>();

			int nColPos = neuronNumber % getColCount();
			int nRowPos = neuronNumber / getColCount();

			int r = (int)Math.floor(radius + 1);

			//runs through rows
			int higherLevel = nColPos - r;
			if(higherLevel < 0)
			{
				higherLevel = 0;
			}
			int lowerLevel = higherLevel + 2*r;
			if (lowerLevel > getRowCount())
			{
				lowerLevel = getRowCount();
			}

			//runs through columns
			int left = nRowPos - r;
			if (left < 0)
			{
				left = 0;
			}
			int right = left + 2*r;
			if (right > getColCount())
			{
				right = getColCount();
			}


			for (int col = higherLevel; col < lowerLevel; col++)
			{
				for (int row = left; row < right; row++)
				{
					double currDist = GetDistance(col, row, nColPos, nRowPos);
					if (currDist < radius)
					{
						result.put(row * getColCount() + col, currDist);
					}
				}
			}

			return result;
		}

		public final java.util.HashMap<Integer, Double> Slow_GetNeuronsInRadius(int neuronNumber, double radius)
		{
			int neuronsCount = getRowCount() * getColCount();

			int neuronColPos = neuronNumber % getColCount();
			int neuronRowPos = neuronNumber / getColCount();

			java.util.HashMap<Integer, Double> result = new java.util.HashMap<Integer, Double>();

			for (int col = 0; col < getColCount(); col++)
			{
				for (int row = 0; row < getRowCount(); row++)
				{
					double currDist = GetDistance(col, row, neuronColPos, neuronRowPos);
					if (currDist < radius)
					{
						result.put(row * getColCount() + col, currDist);
					}
				}
			}

			return result;
		}


		private double GetDistance(int col, int row, int neuronColPos, int neuronRowPos)
		{
			int xD = (neuronColPos - col);
			int yD = (neuronRowPos - row);
			return Math.sqrt(xD * xD + yD * yD);
		}
	}